thunder/transforms/cudagraph.py

```diff
@@ -34,105 +34,155 @@ def extract_descriptor(arg):
+
+    def make_static_inputs_mask(self, fn_name, *args):
+        _, static_inputs_mask = self.python_callables[fn_name]
+        if static_inputs_mask is None:
+            static_inputs_mask = tuple(isinstance(arg, torch.nn.Parameter) for arg in args)
+        return static_inputs_mask
 
-    # Warmup
-    torch.cuda.synchronize()
-    stream = torch.cuda.Stream()
-    stream.wait_stream(torch.cuda.current_stream())
+    def make_cache_key(self, fn_name, *args):
+        # if args_descriptor included torch.nn.Parameter-ness, we would
+        # could use the static_inputs_mask or None as the key.
+        return (fn_name, self.make_static_inputs_mask(fn_name, *args), to_arg_descriptor(*args))
 
-    with torch.cuda.stream(stream):
-        static_inputs = tuple(
-            get_static_buffer(arg) if not is_static else arg for arg, is_static in zip(args, static_args_mask)
-        )
-        for _ in range(3):
-            fn(*static_inputs)
+    def call_cuda_graph(self, fn_name, *args):
+        fn, num_static_inputs = self.python_callables[fn_name]
 
-    stream.synchronize()
-    torch.cuda.current_stream().wait_stream(stream)
-    torch.cuda.synchronize()
+        cache_key = self.make_cache_key(fn_name, *args)
 
-    # Record
-    # NOTE: We are using default private pool here, but it is possibly better to
-    # use a custom pool for better memory management. See CUDA Graphs Tree in
-    # PyTorch's Inductor: torch/_inductor/cudagraph_trees.py
-    # Design doc: https://docs.google.com/document/d/1ZrxLGWz7T45MSX6gPsL6Ln4t0eZCSfWewtJ_qLd_D0E/view
-    graph = torch.cuda.CUDAGraph()
-    with torch.cuda.graph(graph, stream=stream):
-        static_outputs = fn(*static_inputs)
+        cache_entry = self.cuda_graph_cache.get(cache_key)
+        if cache_entry is None:
+            static_inputs_mask = self.make_static_inputs_mask(fn_name, *args)
+            cache_entry = self.build_cuda_graph(fn, args, static_inputs_mask)
+            self.cuda_graph_cache[cache_key] = cache_entry
 
-    return graph, static_inputs, static_outputs
+        graph, static_inputs, static_outputs = cache_entry
 
+        for static_input, arg in utils.safe_zip(static_inputs, args):
+            if id(static_input) != id(arg) and isinstance(static_input, torch.Tensor) and isinstance(arg, torch.Tensor):
+                static_input.copy_(arg)
 
-class CUDAGraphCallable:
-    def __init__(self, fn: Callable, num_static_inputs: None | int = None):
-        self.fn = fn
-        self.num_static_inputs = num_static_inputs
+        graph.replay()
+        return static_outputs
 
-    def __call__(self, *args):
-        if self.num_static_inputs is not None:
-            static_inputs_mask = (True,) * self.num_static_inputs + (False,) * (len(args) - self.num_static_inputs)
-        else:
-            static_inputs_mask = tuple(isinstance(arg, torch.nn.Parameter) for arg in args)
+    def make_python_callable_from_symbols(
```

Comment:
This looks like something we could have in a transform building block library in the future

Reply:
Indeed! The question is whether we should build the region before...